### ISSUE TITLE
Coalesce subscriptions to the Kubernetes EndPoints API

### ISSUE THREAD

**Issue by briansmith** [runconduit/conduit#334]
In #155 @seanmonstar noted that we don't coalesce subscriptions to the EndPoints service, so there will be one EndPoints subscription per (proxy instance, service name). This seems very heavyweight. In particular, it seems worse in terms of load on the EndPoints server than what kube-proxy does.

We should, at least, change this so that conduit doesn't increase the load on the Kubernetes master API EndPoints server beyond what the load would be prior to installing Conduit, even when there are many proxies.

Ideally, Conduit would significantly *reduce* the load on the EndPoints server.

I looked at how kube-proxy is implemented and it watches for all EndPoints changes and keeps its own copy of the service name -> endpoint mapping in memory. See
* https://github.com/kubernetes/kubernetes/blob/cd2692c49d63891ac3c30b416f59df8e3ffcb935/cmd/kube-proxy/app/server.go#L538-L544
* https://github.com/kubernetes/kubernetes/blob/dd9981d038012c120525c9e6df98b3beb3ef19e1/pkg/proxy/config/config.go#L75

I think the Destination service's server should do similar.

Note that I only talked about the EndPoints service above, but the same concerns apply to the subscriptions to the Services service.

/cc @adleong

**Comment by seanmonstar** [comment]
Doing some more reading through controller code, my initial assessment could have been very wrong. It kind of does look like there might be coalescing into a single watcher or informer in https://github.com/runconduit/conduit/blob/master/controller/k8s/endpoints.go, but I don't grasp even 2% of it all, yet.

**Comment by adleong** [comment]
What evidence is there that the destination service is not coalescing subscriptions to the EndPoints service?  Are you observing too many active connections to the kubernetes API from the controller pod?

**Comment by seanmonstar** [comment]
I have no evidence, I mistook the subscriptions to be grpc streams to kubernetes, but instead, it looks like there is actually just 1 and subscriptions are our own app specific thing wrapping the k8s connection.

Excuse the libel.

**Comment by briansmith** [comment]
> What evidence is there that the destination service is not coalescing subscriptions to the EndPoints service? Are you observing too many active connections to the kubernetes API from the controller pod?

I misunderstood the code similar to @seanmonstar, I think. In particular, I misunderstood what this code is doing:

```go
func (s *server) resolveKubernetesService(id string, port int, stream pb.Destination_GetServer) error {
	listener := endpointListener{stream: stream}
	s.endpoints.Subscribe(id, uint32(port), listener)
	<-stream.Context().Done()
	s.endpoints.Unsubscribe(id, uint32(port), listener)
	return nil
}
```

Where's the best place to read a summary of the design of the Destination service server?

**Comment by klingerf** [comment]
This was fixed by #1079.
